Source/Core/AsyncQueue.cs

```diff
@@ -0,0 +1,88 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Boogie;
+
+/// <summary>
+/// A queue from which items can be asynchronously dequeued before adding them. For example:
+///
+/// var queue = new AsyncQueue<int>()
+/// var itemTask = queue.Dequeue();
+/// queue.Enqueue(3);
+/// Assert.AreEqual(3, await itemTask);
+///
+/// All methods are thread-safe.
+/// Also supports pushing elements to the front of the queue.
+/// </summary>
+public class AsyncQueue<T>
+{
+  private readonly object myLock = new();
+  // At all times, either items or customers is empty.
+  private readonly LinkedList<T> items = new();
+  private readonly Queue<TaskCompletionSource<T>> customers = new();
+
+  public void Enqueue(T value)
+  {
+    lock (this) {
+      if (TryEnqueue(value))
+      {
+        return;
+      }
+
+      items.AddLast(value);
+    }
+  }
+
+  private bool TryEnqueue(T value)
+  {
+    while (customers.TryDequeue(out var customer)) {
```

Comment:
Here is a potential deadlock or at least slow down
https://devblogs.microsoft.com/premier-developer/the-danger-of-taskcompletionsourcet-class/
In short, `customer.TrySetResult` will continue to execute all continuations synchronously, preventing other threads to access this queue.

The solution to that is to lock only when dequeuing customers, not on `customer.TrySetResult(value)`

Possible solution
1) Don't lock around `TryEnqueue`
2) 
```
var consumer;
var dequeued = true;
while(dequeued) {
  lock(lock) {
    dequeued = customers.TryDequeue(out customer)
  }
  if(dequeued) {
    if (customer.TrySetResult(value)) {
      return true;
    }
  }
}
```

Reply:
This problem is handled on line 74, and the test `EnqueueReturnsBeforeCompletingDequeueTask` tests against this deadlock

Reply:
Great job.
Just a small question then:
`ContinueWith` will always return a task that succeeds, it does not propagate exceptions like the fact that t was canceled. Is that intended?